hazelcast/src/main/java/com/hazelcast/internal/partition/ChunkSuppliers.java

```diff
@@ -0,0 +1,150 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.partition;
+
+import com.hazelcast.internal.util.CollectionUtil;
+import com.hazelcast.spi.impl.operationservice.Operation;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.BooleanSupplier;
+import java.util.function.Supplier;
+
+import static com.hazelcast.internal.util.Preconditions.checkNoNullInside;
+import static com.hazelcast.internal.util.Preconditions.checkTrue;
+import static com.hazelcast.internal.util.Preconditions.isNotNull;
+
+/**
+ * Helper methods to create various {@link ChunkSupplier} objects
+ */
+public final class ChunkSuppliers {
+
+    private ChunkSuppliers() {
+    }
+
+    /**
+     * Intended use cases:
+     * <ul>
+     * <li>When migration data fits in one single chunk</li>
+     * <li>When a service doesn't support data migration in chunks</li>
+     * </ul>
+     */
+    public static ChunkSupplier newSingleChunkSupplier(Supplier<Operation> operationSupplier) {
+        isNotNull(operationSupplier, "operationSupplier");
+
+        return new SingleChunkSupplier(operationSupplier);
+    }
+
+    private static final class SingleChunkSupplier implements ChunkSupplier {
+
+        private final Supplier<Operation> operationSupplier;
+
+        private boolean hasMoreChunks = true;
+
+        private SingleChunkSupplier(Supplier<Operation> operationSupplier) {
+            this.operationSupplier = operationSupplier;
+        }
+
+        @Nullable
+        @Override
+        public Operation next() {
+            if (!hasMoreChunks) {
+                throw new NoSuchElementException();
+            }
+            Operation operation = operationSupplier.get();
+            hasMoreChunks = false;
+            return operation;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return hasMoreChunks;
+        }
+
+        @Override
+        public String toString() {
+            return "SingleChunkSupplier{"
+                    + "operationSupplier=" + operationSupplier
+                    + ", hasMoreChunks=" + hasMoreChunks
+                    + '}';
+        }
+    }
+
+    /**
+     * @param chain chain of {@link ChunkSupplier}
+     * @return wrapper {@link ChunkSupplier}
+     */
+    public static ChunkSupplier newChainedChunkSupplier(List<ChunkSupplier> chain) {
+        isNotNull(chain, "chain");
+        checkTrue(CollectionUtil.isNotEmpty(chain), "chain cannot be an empty list");
+        checkNoNullInside(chain, "chain cannot have null value inside");
+
+        return new ChainedChunkSupplier(chain);
+    }
+
+    private static final class ChainedChunkSupplier implements ChunkSupplier {
+
+        private final int length;
+        private final List<ChunkSupplier> chain;
+
+        private ChainedChunkSupplier(List<ChunkSupplier> chain) {
+            this.chain = chain;
+            this.length = chain.size();
```

Comment:
very minor : what if we use queue rather than list for the `chain`, in the upper layer linkedlist can be passed rather than arraylist. It will also help with redundant iterations in the `hasNext()` and `next()` method, wdyt?

Reply:
list seems to fit better here than queue since we may revisit elements, for now lets keep it as list.